lib/catalog.ex

```diff
@@ -0,0 +1,112 @@
+defmodule Catalog do
+  use GenServer
+
+  alias Catalog.{Bucket, Document}
+
+  @type t :: %__MODULE__{
+          ets: :ets.tid(),
+          loaded?: boolean()
+        }
+
+  @enforce_keys [:ets, :loaded?]
+
+  defstruct ets: nil,
+            loaded?: false
+
+  # Client
+
+  @spec start_link(keyword()) :: GenServer.on_start()
+  def start_link(opts) do
+    name = Keyword.get(opts, :name, __MODULE__)
+
+    GenServer.start_link(__MODULE__, opts, name: name)
+  end
+
+  @spec lookup(String.t() | nil, String.t() | nil, Date.t() | nil) ::
+          {:ok, [Document.t()]} | {:error, String.t()}
+  @spec lookup(GenServer.server(), String.t() | nil, String.t() | nil, Date.t() | nil) ::
+          {:ok, [Document.t()]} | {:error, String.t()}
+  def lookup(pid \\ __MODULE__, last_name, first_name, date_of_birth)
+
+  def lookup(_, nil, nil, nil),
+    do: {:error, "Bad request. You must include at least one field to search on."}
+
+  def lookup(pid, last_name, first_name, date_of_birth) do
+    GenServer.call(pid, {:lookup, last_name, first_name, date_of_birth})
+  end
+
+  # Server
+
+  @impl GenServer
+  @spec init(keyword()) :: {:ok, t(), {:continue, {:load_documents, keyword()}}}
+  def init(opts) do
+    ets = :ets.new(__MODULE__, [:set, :protected, {:read_concurrency, true}])
+
+    {:ok, %__MODULE__{ets: ets, loaded?: false}, {:continue, {:load_documents, opts}}}
+  end
+
+  @impl GenServer
+  @spec handle_continue({:load_documents, keyword()}, t()) :: {:noreply, t()}
+  def handle_continue({:load_documents, opts}, %__MODULE__{ets: ets, loaded?: false} = state) do
+    documents_fn = Keyword.get(opts, :documents_fn, &all_documents/0)
+
+    for document <- documents_fn.() do
+      _ =
+        :ets.insert(
+          ets,
+          {Document.key(document), document.last_name, document.first_name,
+           document.date_of_birth, document}
+        )
+    end
+
+    {:noreply, %__MODULE__{state | loaded?: true}}
+  end
+
+  @impl GenServer
+  def handle_call(
+        {:lookup, last_name, first_name, date_of_birth},
+        _from,
+        %__MODULE__{ets: ets} = state
+      ) do
+    results =
+      ets
+      |> :ets.match_object(
+        {:_, search_term(last_name), search_term(first_name), search_term(date_of_birth), :_}
+      )
+      |> Enum.map(fn {_key, _last, _first, _dob, document} -> document end)
```

Comment:
If all you need is part of the row, look at `:ets.match` where you can only return part of the record:
```elixir
ets
|> :ets.match({:_, search_term(last_name), search_term(first_name), search_term(date_of_birth), :"$1"})
|> List.flatten()
```